STYLE_GUIDE.md

```diff
@@ -1,3 +1,66 @@
 # Fortran stdlib Style Guide
 
-This document will describe the code style to use.
+> [A]utomate and codify as much as you possibly can while remembering that the human touch is still necessary to praise
+> and be kind to your contributors.
+> Let robots handle your project’s pedantry and humans handle your project’s empathy.
+
+-- @mikemcquaid, [Homebrew] project leader<sup>[1]</sup>
+
+[1]: https://mikemcquaid.com/2018/06/05/robot-pedantry-human-empathy/
+[Homebrew]: https://brew.sh
+
+## File naming conventions
+
+- Source files should contain at most one `program` or `module`/`submodule`.
+- The filename should match the program or module name and have the file extension `.f90` or `.F90` if preprocessing is required.
+- If the interface and implementation is split using submodules the implementation submodule file should have the same name as the interface (parent) module but end in `_implementation`. E.g., `string_class.f90` and `string_class_implementation.f90`
+- Tests should be added in the `tests` subdirectory and have the same name as the module they are testing with the `test_` prefix added. E.g., `string_class.f90` and `tests/test_string_class.f90`
+
+## Indentation & whitespace
+
+By setting and following a convention for indentation and whitespace, code reviews and git-diffs can
+focus on the semantics of the proposed changes rather than style and formatting.
+
+* __TABs vs spaces__: __Spaces__: TABs should never be used for indentation
+* __Trailing whitespace__: All trailing whitespace must be removed
+* __End of File (EOF)__: The file must end in a single new-line character
+* __Line length__: __132__: The Fortran 90 (and later) standard mandates lines be no longer than 132 characters
```

Comment:
This whole section seems overly technical, rather than part of the guide. I suggest we extract the actionable suggestions and list them in plain language, for example:

* The body of every Fortran construct should be indented by 4 spaces
* Line length should be limited to 80 characters, but must not exceed 132.

If trailing whitespaces and EOL and EOF stuff will be handled by CI, no need to mention it in the guide.

Reply:
Regarding 80 characters, I follow very strictly 80 character limit (and frequently have several terminals side by side on a screen, which the 80 characters allow), but not everybody does, and so I like the way you phrased it as `Line length should be limited to 80 characters and must not exceed 132.`, to recommend to try to keep it under 80, but if it can't be done, then one can go over. @zbeekman is that an acceptable compromise for you?

Reply:
@certik yes, I like the way I phrased the line length stuff to address line length. Keeping it less than 80 is ideal, but under 132---to provide an absolute limit somewhere, and conform to the standard---is a good compromise IMO. Especially if you have 4 spaces of indentation.

@milancurcic : Well, when you say:

> If trailing whitespaces and EOL and EOF stuff will be handled by CI, no need to mention it in the guide.

what are you envisioning?

If contributors aren't notified ahead of time about CI checks it could be frustrating to learn after you've committed code that you need to make stylistic changes. Or are you suggesting that we have CI automatically alter and commit changes to contributed code so that it would conform with the style?

Reply:
I've done my best to simplify and cleanup this section using more plain language, and respond to your suggestions.

@milancurcic I've included some stuff about tabs and trailing white space because I think it's worthwhile to be up-front about it, and don't envision a CI system commiting changes on a users behalf.

Reply:
> Or are you suggesting that we have CI automatically alter and commit changes to contributed code so that it would conform with the style?

Sorry, indeed I assumed this would be done automatically via CI. I agree, if the developer is to worry about these things, we should state it in the style guide.

Reply:
We can setup pre-commit hooks to automatically fix issues like this and do auto-formatting, but it's hard to ensure that people are always using them.

In CI, I don't think it makes sense (at least not right now) to have CI authoring commits on the users behalf.